<!DOCTYPE html>
<html lang="en-us">

  <head>

    <meta charset="UTF-8">
   <title>RABV-GLUE by giffordlabcvr: Comparative analysis of rabies virus (RABV) genomes</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="website/assets/stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="website/assets/stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="website/assets/stylesheets/github-light.css" media="screen">

  </head>

  <body>

    <section class="page-header">
      <h1 class="project-name">RABV-GLUE</h1>
      
      <h2 class="project-tagline">Comparative analysis of rabies virus (RABV) genomes</h2>


      <a href="https://github.com/giffordlabcvr/RABV-GLUE/zipball/master" class="btn">Download</a>
      <a href="https://github.com/giffordlabcvr/RABV-GLUE/tree/master" class="btn">GitHub</a>
    </section>

    <section class="main-content">

	   <h3>
		 <a id="overview" class="anchor" href="#overview" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Overview</strong>
	   </h3>

		 <p>
		   This is <b>RABV-GLUE</b>, a GLUE
		   project designed to support comparative genomic and evolutionary analysis of
		   <a target="_blank" href="https://www.cdc.gov/rabies/about.html"><strong>rabies virus (RABV)</strong></a>.
		   <b>RABV-GLUE</b> contains aligned, annotated RABV genome sequence data.
		   
		 </p>


	   <h3>
	   <a id="glueProject" class="anchor" href="#glueProject" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>GLUE project</strong>
	   </h3>

      	   
	   <p>
	
         On computers with <a href="http://glue-tools.cvr.gla.ac.uk/#/installation">GLUE installed</a>,
         the <b>RABV-GLUE</b> project can be instantiated by navigating to the project folder,
         initiating GLUE, and issuing the following command in the GLUE shell:
         
  <pre>
  <code>Mode path: /
  GLUE> run file cressProject.glue</code></pre>
		 
	   </p>



       <br>
	   <h3>
	   <a id="WhatGlueProject" class="anchor" href="#glueProject" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>What is a GLUE project?</strong>
	   </h3>
       <hr>

	   <p>
		 <a target="_blank" href="http://glue-tools.cvr.gla.ac.uk/"><b>GLUE</b></a> is an open, integrated
		 software toolkit that provides functionality for storage and interpretation of
		 sequence data. 
	   </p>

	   <p>
		 <b>GLUE</b> supports the development of <b>“projects”</b> containing the data items
		 required for comparative genomic analysis 
		 (e.g.  sequences, multiple sequence alignments, genome feature annotations,
		 and other sequence-associated data). 
	   </p>


        <br>
        <p><img src="./website/assets/images/glue-web.jpg" alt="GLUE web" /></p>
        <br>


	   <p>
		 Projects are loaded into the <b>GLUE "engine"</b>, creating a <b>relational database</b>
		 that represents the semantic relationships between data items.
		 This provides a robust foundation for the implementation of <b>systematic
		 comparative analyses</b> and the development of <b>sequence-based resources</b>. 
	   </p>

	   <p>		 
		  The core schema of this database can be extended to accommodate
		  the idiosyncrasies of different projects, and
		  GLUE provides a scripting layer (based on JavaScript) for developing custom
		  analysis tools. 
	   </p>

	   <p>
		  Hosting of GLUE projects in an online version control system (e.g. GitHub) provides
		  a mechanism for their stable, collaborative development.
	   </p>
	   
	   

		   
		 <p>
		   Some examples of '<b>sequence-based resources</b>' built for viruses using GLUE include:

		 </p>
		   
		   

		 <p>

		 <ul>
		 		 
		   <li> 
			 <a target="_blank" href="http://cov-glue.cvr.gla.ac.uk/"><b>COV-GLUE</b></a>: 
			 A GLUE resource for tracking genetic variation in SARS-COV2.
			 CoV-GLUE contains a database of amino acid replacements, insertions and
			 deletions which have been observed in GISAID hCoV-19 sequences sampled from the pandemic
		   </li>
		   <br>
		   
		   <li>
			 <a target="_blank" href="http://hcv-glue.cvr.gla.ac.uk/"><b>HCV-GLUE</b></a>:
			 This GLUE resource aims to support analysis of drug resistance and vaccine
			 escape in hepatitis C virus (HCV).
			 A database of HCV sequences and metadata from NCBI, updated daily and arranged
			 into clades (genotypes, subtypes). As well as pre-built multiple-sequence
			 alignments of NCBI sequences, it includes an analysis tool providing genotyping,
			 drug resistance analysis and visualisation of submitted FASTA sequences.
           </li>
	   
		 </ul>

		   
		 </p>
	   


	   <br>
	   <h3>
	   <a id="WhyGlueProject" class="anchor" href="#glueProject" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>What does building the Flaviviridae-GLUE project offer?</strong>
	   </h3>
	   <hr>

		 
 
	   <p>
		 
		  Flaviviridae-GLUE offers a number of advantages for performing comparative sequence
		  analysis of flaviviruses:


		  <ol>

			<li> <strong>Reproducibility</strong>. 
			For many reasons, bioinformatics analyses are notoriously difficult to reproduce.
			The GLUE framework supports the implementation of fully reproducible 
			comparative genomics through the introduction of data standards and the use 
			of a relational database to capture the semantic links between data items.
			</li>
			<br>
		  
			<li> <strong>Reusable data objects and analysis logic</strong>. 
			For many - if not most - comparative genomic analyses, data preparation is nine
			tenths of the battle. The GLUE framework has been designed to ensure that 
			work spent preparing high-value data items such as multiple sequence alignments
			need only be performed once. Hosting of GLUE projects in an online version control 
			system such as GitHub allows for collaborative management of important data items
			and community testing of hypotheses.
			</li>
			<br>


			<li> <strong>Validation</strong>. 
			Building GLUE projects entails mapping the semantic links between data items
			(e.g. sequences, tabular data, multiple sequence alignments).
			This process provides an opportunity 
			for cross-validation, and thereby enforces a high level of data integrity.
			</li>
			<br>
		
			<li> <strong>Standardisation of the genomic co-ordinate space</strong>. GLUE
			projects allow all sequences to utilise the coordinate space of a chosen 
			reference sequence. Contingencies associated with insertions and deletions
			(indels) are handled in a systematic way.
			</li>
			<br>
		  
			<li> <strong>Predefined, fully annotated reference sequences</strong>:
			This project includes fully-annotated reference sequences for major lineages
			within the <i>Hepadnaviridae</i> family.
			</li>
			<br>
		  
			<li> <strong>Alignment trees</strong>:
			GLUE allows linking of alignments constructed at distinct taxonomic levels
			via an ""<b>alignment tree</b>"
			data structure. In the alignment tree, each alignment
			is constrained to a standard reference sequence, thus all multiple sequence
			alignments are linked to one another via a standardised coordinate system.
			</li>
			
		  </ol>


	   </p>


	   <h3>
		 <a id="contributors" class="anchor" href="#contributors" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Contributors</strong>
	   </h3>
	   
	   <p>
		  Robert J. Gifford (<a href="mailto:robert.gifford@glasgow.ac.uk">robert.gifford@glasgow.ac.uk</a>)	 
	   </p>

	   
	   <p>
		  Kathryn Campbell
	   </p>
	   
	   <p>
		  Kirstyn Brunker
	   </p>




	   <h3>
		 <a id="Publications" class="anchor" href="#Publications" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>Related Publications</strong>
	   </h3>

 
	   <p>
		 
		  Brunker K, Gurdeep G, Thumbi SM, Lushasi K, Lugelo A, Czupryna AM, Ade F, Wambura G, Chuchu V, Steenson R,
		  Ngeleja C, Bautista C, Gomez MRR, Chu MYJV, Manalo DL, Miranda MA,  Kamat KT, Rysava K, Espineda J,
		  Silo EAV, Aringo AM, Bernales RP, Adonay FF, Tildesley MJ, Marston DA, Jennings DL, Zhu W, Meredith LW, 
		  Gifford RJ, B. Singer JB, Muturi M, Mwatondo A, Biek R, and K Hampson
		  <b>(2019)</b>
		  <br>
		  Rapid in-country sequencing of virus genomes to inform rabies elimination programmes.
		  <i>Wellcome Open Research</i>  
		  <a href="https://wellcomeopenresearch.org/articles/5-3/v1" target="_blank">[view]</a>
		  <br>
		  <br>

		  Singer JB, Thomson EC, McLauchlan J, Hughes J, and RJ Gifford
		  <b>(2018)</b>
		  <br>
		  GLUE: A flexible software system for virus sequence data.
		  <br>
		  <i>BMC Bioinformatics</i>
		  <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2459-9" target="_blank">[view]</a> 
		  <br>

 		</p>


	   <h3>
		 <a id="license" class="anchor" href="#license" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a><strong>License</strong>
	   </h3>

	   <p>
		 This project is licensed under the <a href="https://www.gnu.org/licenses/agpl-3.0.en.html">GNU Affero General Public License v. 3.0</a>.
	   </p>




	 <footer class="site-footer">
	   <span class="site-footer-owner"><a href="https://github.com/giffordlabcvr/RABV-GLUE">RABV-GLUE</a> is maintained by <a href="https://github.com/giffordlabcvr">giffordlabcvr</a>.</span>

	   <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
	 </footer>

    </section>


  </body>
  
</html>
